(docs) Add Contributing Guidelines (#2760)
highlightjs/highlight.js: CONTRIBUTING.md

# Contributing

**Table of Contents**
1. [Welcome](#welcome) (2nd draft)
2. [Philosophy](#philosophy) (TBD)
3. [Reporting an Issue](#reporting-an-issue) (TBD)
4. [Fixing an Issue (PRs)](#fixing-an-issue-prs) (2nd draft)
5. [Setup and Run](#setup-and-run) (TBD)

## Welcome
Hi, and welcome to Highlight.js. We're a code-highlighting engine etc., but you probably know that. You want to contribute, and this document will be your guide.

> There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or creating [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).

Reviewer: >adding [a 3rd party grammar

There is also this page: https://highlightjs.readthedocs.io/en/latest/language-contribution.html
which is broader (it includes the link to the 3RD_PARTY_QUICK_START.md)

Reviewer 2: Yes, there is already talk of merging them long-term in which case this obviously should link to the consolidated document.  This could mention both if it needed to until then - is that what you're suggesting?

> There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or creating [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).

Reviewer 2: Probably not, but if we aren't disallowing are you suggest we merely discourage by not mentioning it or actively disallow?  Or find some way to get them maintainers on the hook for maintaining themes?

> There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or creating [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).

Author: How about "working on a theme" instead of "creating a theme", and linking to a list of existing themes as well?

> There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or creating [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).

Author: ```suggestion
There are many ways to contribute to Highlight.js. Reporting issues is super useful. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html), adding [a 3rd party grammar](https://github.com/highlightjs/highlight.js/blob/master/extra/3RD_PARTY_QUICK_START.md), or developing [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html).
```

> In terms of getting help, GitHub issues are the primary method of communication. We're exploring the idea of creating a Gitter chat room.

Reviewer 2: I guess it doesn't add much.

> In terms of getting help, GitHub issues are the primary method of communication. We're exploring the idea of creating a Gitter chat room.

Author: ```suggestion
If you need support , [open a GitHub issue](https://github.com/highlightjs/highlight.js/issues/new/choose).
```

> ## Philosophy

Author: A bit long, imho. People need to actually _read_ it 😁.

Reviewer 2: Really?  I didn't find it that along... I can take a second look... did you imagine just a bullet point list or?   What length do you imagine is "perfect" for such a thing?

Author: multiple sections is fine. I'm mainly worried I usually don't read documents on the web word-for-word, so it should be a easy to scan.

Highlight.js makes a strong effort to keep the core engine, it's usage, and it's feature set clean and simple.  We aim to support the happy path (the 90%), not the often innumerable edge cases (the last 10%).  When you're on the happy path everything "just works" with a single line of code.  We do our best to allow those familiar with JavaScript to easily handle the remaining 10% via small extensions or plug-ins.

> Feature requests are always welcome but many times the answer is:
> 
> > That's a great idea, you should write a plug-in for that! It doesn't belong in the core library and it's probably something you could do with only a small amount of code.
> 
> If you'd like to contribute a feature you should always first open an issue to discuss if the feature belongs in core vs a plug-in.  Often this is a great way to get tips for implemention as well or links to prior disussions with context, etc.

Author: This gives the reader a good overview of features. Maybe give it a heading, so we can link to it elsewhere, eg. `#prerequisites`. This would mean the next section needs a heading too, so it's not under `#feature-requests` (faq?)
```suggestion
### Feature Requests
Feature requests are always welcome, but many times the answer is:

> That's a great idea, you should write a plug-in for that! It doesn't belong in the core library and it's probably something you could do with only a small amount of code.

If you'd like to contribute a feature you should always first open an issue to discuss if the feature belongs in core vs a plug-in.  Often this is a great way to get tips for implemention as well or links to prior disussions with context, etc.
```

Author: Actually, `#reporting-an-issue` looks empty. Maye move the feature stuff there.

Reviewer 2: I think of feature requests and issues as entirely different things.

Reviewer 2: I think feature requests makes a great section.  I mean we have a table of contents so as long as this doesn't get too long I'm ok with it having some meat.

Some bullet points:

> - Our [lack of line numbers is a feature](https://highlightjs.readthedocs.io/en/latest/line-numbers.html). The simple case isn't hard to add though with a little custom JavaScript.

Reviewer: Found a reference to the google group on that page... Someone monitors it?
And shouldn't we point users to https://github.com/taufik-nurrohman/highlight.ln.js?

Reviewer 2: We could very briefly, though I think we need another separate place for "Awesome HLJS" stuff in general.

## Fixing an Issue (PRs)
If you're new to contributing to open-source, have a look at [this GitHub Guide](https://guides.github.com/activities/forking). It explains the general process of GitHub collaboration.

If you feel comfortable with our [prerequisites](#prerequisites), you can grab any issue [marked `beginner friendly`](https://github.com/highlightjs/highlight.js/issues?q=is%3Aopen+is%3Aissue+label%3A%22beginner+friendly%22).

### Rules
Don't worry; these rules are pretty standard.

Please open an issue before your PR, so we can explore and discuss the topic at hand. Your time is important, and we need to be sure it's well-spent.

**Before** you start coding, keep these tips in mind:
- You should usually add markup tests (ie. whenever you've made a significant grammar change or fixed a bug)
- Change only what needs to be changed; don't re-lint/rewrite whole files when fixing bugs
- Linting needs a dedicated commit

**When you're ready** to open a PR, we have some housekeeping tasks for you.
- Find and link to the relevant issue
- Add a changelog entry to [CHANGES.md](https://github.com/highlightjs/highlight.js/blob/master/CHANGES.md)

> - If necessary, add your name to [AUTHORS.txt](https://github.com/highlightjs/highlight.js/blob/master/AUTHORS.txt) under Contributors

Reviewer 2: Is this redundant now that it's in our PR template or is this just good reinforcing?

Author: I meant it as reinforcement. We could remove it, if you want.

Reviewer 2: Feels a bit redundant... (the list)... It could just say "when you open a PR we'll have a small checklist waiting for you... it's easy".  No need to repeat.

## Setup and Run

> You'll of course need Node.js and npm or yarn.

Reviewer 2: Do we need to link to install node or something here?

Author: It wouldn't hurt. Although, npm comes with Node.js iirc.

Reviewer 2: Really, has the world just decided npm is dead or is that more of a personal choice :)  Or does node include NPM now out of the box?

Author: Node includes npm out-of-the-box.

Often when contributing a PR (that doesn't make any specific changes to browser
features) it's more than sufficient to build and test only the Node.js build.
Our CI process will guarantee that the browser build is still green.

**Should we shown yarn examples also?**

```
npm run build
npm run test
```

The browser library must be built and tested separately:

```
npm run build-browser
npm run test-browser
```

> **Where should we mention the developer tool? ./tools/developer.html**

Author: It's already in the "Building & Testing" docs linked below, which should be enough.

Author: actually, I'll reference it explicitly.

The "run" here made me think of that because otherwise there isn't really
much to "run" (other than tests).

> For more details see [Building and Testing](https://highlightjs.readthedocs.io/en/latest/building-testing.html).

Author: Tiny aesthetics.
```suggestion
For more details see the docs on [Building and Testing](https://highlightjs.readthedocs.io/en/latest/building-testing.html).
```

Reviewer 2: Ok let's not run these tiny one offs by me... please fix grammar, tweak typos, rewrite small sections as necessary... take the initiative, whatever.  I'm trying to focus more on the 1 mile high view. :-)

### Prerequisites
Are you someone who...
- knows JavaScript?
- has a working familiarity with Regex? (you won't get too far without this)
- perhaps has some rough familiarity with the idea of parsing text in general?
- is willing to read our grammar documentation and look at other grammars?

> 

Reviewer 2: We need to find a way to expand this to make it clear there are multiple potential things to work on... someone could submit a new theme without knowing any JS or regex, etc... When I wrote this i was primarily thinking of fixing grammar bugs, etc (our most common type of issue).

# Contributing
**Table of Contents**

> - [Getting Started](#getting-started) (1st draft)

Reviewer 2: "Getting Started" vs "Setup and Run" is confusing to me... is this really more of "Welcome"?  Maybe not, but right now the distinction isn't clear. :-)

Author: "Welcome" sounds right.

- [Philosophy](#philosophy) (TBD)
- [Reporting an Issue](#reporting-an-issue) (TBD)
- [Pull Requests](#pull-requests) (1st draft)

> 	-  _Should this be renamed to "fixing an issue"?_

Reviewer 2: No objection... `Fixing an Issue (via Pull Request)`?

Author: That sounds too verbose to me.

Author: `Fixing an Issue (PR)`?

	-  Link to GitHub guide for PRs

> 	-  What issues are good to start with? (Eg. `feature`, `bug`, etc. )

Reviewer 2: "help welcome" but more specifiically "beginner friendly"... although somewhere we need to disclaim what we mean by "beginner".

Author: To go down that Rabbit hole... do we mean beginner to _the project_ or to _coding_?

Reviewer 2: Ha, perhaps the whole tags are named poorly.  I mean that for someone who:

- knows Javascript
- has a working familiarity with Regex (you won't get too far without this)
- perhaps has some rough familiarity with the idea of parsing text in general
- willingness to read our grammar documentation and look at other grammars

If I think I could possibly fix something in 5-30 minutes I'd consider it potentially beginner friendly.  (ie it's perhaps a simpler rule change that doesn't get into the nuance of our grammars or complex things).  Or there are some issues only involving HTML/CSS/themes/docs... I would tag those beginner friendly.  I guess I meant "to the project"?

I'm open to renaming the tags or coming up with a different classification system if it would be helpful.

I slap most things that don't affect the core parser with "help wanted"... things that I'd like to do myself or where I think someone might be more of a hindrance than help I just leave untagged (and someone could always join those discussions).

- [Setup and Run](#setup-and-run) (TBD)

> ## Getting Started

Reviewer 2: Whitespace please! :)

Author: done

> Hi, and welcome to highlight.js. We're a code-highlighting engine etc., but you probably know that. You want to contribute, and this document is the your guide.

Reviewer 2: ```suggestion
Hi, and welcome to Highlight.js. We're a code-highlighting engine etc., but you probably know that. You want to contribute, and this document is the your guide.
```

Correct naming (this applies everywhere).

> There are many ways to contribute to highlight.js. Feel free to work on an open issue. You can also help by writing [a plugin](https://highlightjs.readthedocs.io/en/latest/plugin-api.html) or creating [a theme](https://highlightjs.readthedocs.io/en/latest/style-guide.html). In terms of getting help, GitHub issues are the primary method of communication. We're exploring the idea of creating a Gitter chat room.

Reviewer 2: Isn't reporting an issue also contributing?

Perhaps mention creating a 3rd party grammar here (with link to the quick start for that)?

Author: This https://highlightjs.readthedocs.io/en/latest/language-guide.html? Or the md document?

Reviewer 2: I was thinking of `extra/3RD_PARTY_QUICK_START.md` which was thrown together to answer "I want to make a 3rd party grammar, what do I do?"... Surely it could be improved also. :)

## Philosophy
> A short section on our overall philosophy somewhere would be great also (I'm happy to help there):
> - keep the core parser simple (focus on the 80%)
> - encourage and support a plug-in culture (let others tackle the last 20%)
> - we don't show line numbers on purpose
> - we're more than a keyword highlighter
> - we're not a full language parser
> - auto-detect is not powered by magic unicorn dust
> 
> ([Josh Goebel](https://github.com/highlightjs/highlight.js/issues/2753#issuecomment-709415929))

> ## Reporting an Issue

Reviewer 2: This should definitely link to our great JSFiddle with some instructions on how to use it to provide a test case...

https://jsfiddle.net/ajoshguy/2bmdswn6/

Author: revised

## Pull Requests

> When opening a PR, follow the template.

Reviewer 2: This should probably be two lists... "before you open a PR" and then then "when" list...

"Open an issue first, so we can explore and discuss it"

This is worth a paragraph.  I can flesh it out if you just get something down.

- Add a changelog entry to CHANGES.md
- If necessary, add your name to AUTHORS.txt under Contributors
- You should usually add markup tests (ie. whenever you've made a significant grammar change or fixed a bug)
- Change only what needs to be changed; don't re-lint/rewrite whole files when fixing bugs
- Open an issue first, so we can explore and discuss it
- Linting needs a dedicated commit

## Setup and Run

> (In-depth docs at https://highlightjs.readthedocs.io/)

Reviewer 2: Build and test if probably long enough that's it's ok to have some short summary here (with a link to build and test)... also want to make sure we're not stepping onto the toes of the README, so I think this needs to focus on "I want to setup as a contributor" rather than "how do I use the library on my webpage" (which should be covered well by the readme).

Author: I'm not familiar enough with the library to write this. Do you mind writing it?

Reviewer 2: I just pulled in the TLDR pretty much, thoughts?

Author: I think it's fine.

If you need support, you can [open a GitHub issue](https://github.com/highlightjs/highlight.js/issues/new/choose).

## Philosophy

Highlight.js makes a strong effort to keep the core engine, it's usage, and it's feature set clean and simple.  We aim to support the happy path (the 90%), not the often innumerable edge cases (the last 10%).  When you're on the happy path everything "just works" with a single line of code.  We do our best to allow those familiar with JavaScript to easily handle the remaining 10% via small extensions or plug-ins.

Some bullet points:

- Our [lack of line numbers is a feature](https://highlightjs.readthedocs.io/en/latest/line-numbers.html). It's usually pretty simple to add them yourself ([example](https://github.com/taufik-nurrohman/highlight.ln.js)).
- We [embrace plug-ins and extensions](https://github.com/highlightjs/highlight.js/issues/2225) rather than adding lots of tiny features and configuration options.

> - We are much more than a keyword highlighter. We do make attempts to understand context...

Reviewer 2: ```suggestion
- We are more than a keyword highlighter in that we do make _some_ attempt to understand context...
```

- ...Yet we are not a full parser for any language, we're still just a fancy nested pattern matcher.
- Our [auto-detect is sadly not powered by magic unicorn dust](https://github.com/highlightjs/highlight.js/issues/1213). It's best effort, not best in class.

## Feature Requests
Feature requests are always welcome but many times the answer is:

> That's a great idea, you should write a plug-in for that! It doesn't belong in the core library and it's probably something you could do with only a small amount of code.

If you'd like to contribute a feature you should always first open an issue to discuss if the feature belongs in core vs a plug-in.  Often this is a great way to get tips for implemention as well or links to prior disussions with context, etc.

## Reporting an Issue

> (Misc.)

Reviewer 2: Lets link to the URL for creating new issues here.

Author: ```suggestion
If you find a bug or think of an improvement, feel free to [open an issue](https://github.com/highlightjs/highlight.js/issues/new/choose).
```

- If you found a language highlighting issue, use [this JSFiddle](https://jsfiddle.net/ajoshguy/2bmdswn6/) to create a test case.

## Fixing an Issue (PRs)
If you're new to contributing to open-source, have a look at [this GitHub Guide](https://guides.github.com/activities/forking). It explains the general process of GitHub collaboration.

> If you feel comfortable with our [prerequisites](#prerequisites), you can grab any issue [marked `beginner friendly`](https://github.com/highlightjs/highlight.js/issues?q=is%3Aopen+is%3Aissue+label%3A%22beginner+friendly%22).

Reviewer 2: The label changed to "good first issue".

Author: ```suggestion
If you feel comfortable with our [prerequisites](#prerequisites), you can grab any issue [marked "good first issue"](https://github.com/highlightjs/highlight.js/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22).
```

### Rules
Don't worry; these rules are pretty standard.

Please open an issue before your PR, so we can explore and discuss the topic at hand. Your time is important, and we need to be sure it's well-spent.

**Before** you start coding, keep these tips in mind:
- You should usually add markup tests (ie. whenever you've made a significant grammar change or fixed a bug)
- Change only what needs to be changed; don't re-lint/rewrite whole files when fixing bugs
- Linting needs a dedicated commit

**After** you make your changes, we have some housekeeping tasks for you. See the PR template for details.

## Setup and Run

You'll need [Node.js](https://nodejs.dev/download/), and optionally [yarn](https://yarnpkg.com/).

Often when contributing a PR (that doesn't make any specific changes to browser
features) it's more than sufficient to build and test only the Node.js build.
Our CI process will guarantee that the browser build is still green.

```console

> npm run build

Reviewer 2: Would `build_and_test` be better here?

Author: I think it feels more natural to run two standard npm commands, than to run one non-standard command. At least, for a beginner.

npm run test
```

The browser library must be built and tested separately:

```console

> npm run build-browser

Reviewer 2: `build_and_test_browser` ?

Author: This should be consistent with whatever we do above.

npm run test-browser
```

> There's also a tool for visual testing at [tools/developer.html](https://github.com/highlightjs/highlight.js/blob/master/tools/developer.html)

Reviewer 2: ```suggestion
There's also a tool for visual testing at [tools/developer.html](https://github.com/highlightjs/highlight.js/blob/master/tools/developer.html). It requires a browser build to be present. (ie `npm run build_browser` first)
```

For more details, see [Building and Testing](https://highlightjs.readthedocs.io/en/latest/building-testing.html) in our docs.

### Prerequisites

> Are you someone who...
> - knows JavaScript?
> - has a working familiarity with Regex? (you won't get too far without this)
> - perhaps has some rough familiarity with the idea of parsing text in general?
> - is willing to read our grammar documentation and look at other grammars?
> 

Author: ```suggestion
Are you someone who...
- knows JavaScript?
- has a working familiarity with Regex? (you won't get too far without this)
- perhaps has some rough familiarity with the idea of parsing text in general?
- is willing to read our grammar documentation and look at other grammars?
- wants to help, but doesn't fit any of these requirements?
```
